### PR TITLE
Add mappings

### LINKED ISSUE #99: Implement Piola mappings
These can then be used directly by dolfinx (see FEniCS/ffcx#299)

### LINKED ISSUE #102: Fix mapping name
Should be "identity" as per UFL, not "affine".

### PR DESCRIPTION
**Pull request by mscroggs**
- Added `mapping::type` enum to use instead of string identifier.
- Implemented Piola mappings.
- Will allow us to remove `transform_reference_basis` from ffc and use these mappings directly.

Fixes #102. Fixes #99.

Progress towards #97 (type_to_str is now wrapped)

This PR is coupled with FEniCS/ffcx#304 and FEniCS/dolfinx#1369

### CODE REVIEW

File: cpp/basix.h [hunk 1]
```diff
@@ -17,6 +17,11 @@ void release_element(int handle);
 std::vector<Eigen::ArrayXXd> tabulate(int handle, int nd,
                                       const Eigen::ArrayXXd& x);
 
+/// Apply mapping
```
Comment (garth-wells) [inline cpp/basix.h:20]: What map? I could suggest a name but I don't know from the docstring what it does. How many point does it work on? Can the operation it performs be described?

File: cpp/finite-element.h [hunk 2]
```diff
@@ -215,9 +216,29 @@ class FiniteElement
...
-  const std::string& mapping_name() const;
+  const mapping::type mapping_type() const;
+
+  /// Map a function value from the reference to a cell
+  /// @param reference_data The function value on the reference
+  /// @param J The Jacobian of the mapping
+  /// @param detJ The determinant of the Jacobian of the mapping
+  /// @param K The inverse of the Jacobian of the mapping
+  /// @return The function value on the cell
+  Eigen::ArrayXd apply_mapping(const Eigen::ArrayXd& reference_data,
```
Comment (garth-wells) [inline cpp/finite-element.h:229]: Could we call this `map_push_forward`?

File: cpp/finite-element.h [hunk 3]
```diff
@@ -215,9 +216,29 @@ class FiniteElement
...
+                               const Eigen::MatrixXd& J, double detJ,
+                               const Eigen::MatrixXd& K) const;
+
+  /// Map a function value from a cell to the reference
+  /// @param physical_data The function value on the cell
+  /// @param J The Jacobian of the mapping
+  /// @param detJ The determinant of the Jacobian of the mapping
+  /// @param K The inverse of the Jacobian of the mapping
+  /// @return The function value on the reference
+  Eigen::ArrayXd apply_inverse_mapping(const Eigen::ArrayXd& physical_data,
```
Comment (garth-wells) [inline cpp/finite-element.h:239]: Could we call this `map_pull_back`?

File: cpp/finite-element.h [hunk 4]
```diff
@@ -215,9 +216,29 @@ class FiniteElement
   /// @return The family name
   const std::string& family_name() const;
 
-  /// Get the mapping used for this element
+  /// Get the mapping type used for this element
   /// @return The mapping
-  const std::string& mapping_name() const;
+  const mapping::type mapping_type() const;
+
+  /// Map a function value from the reference to a cell
```
Comment (garth-wells) [inline cpp/finite-element.h:223]: Should this be in the public API header file? (`basix.h`)?

```
/// Map a function value on the reference cell to the physical cell
```
?

**Review by garth-wells — CHANGES_REQUESTED** [review]
I've added a few comments on clarifications. Can look more closely once a few points are clarified.

File: cpp/basix.h (hunk 1 above)
Comment (mscroggs) [reply cpp/basix.h]: I've renamed this to map_push_forward and added a better docstring.

File: cpp/finite-element.h (hunk 4 above)
Comment (mscroggs) [reply cpp/finite-element.h]: It's now in both

**Review by chrisrichardson — APPROVED** [review]

**Comment by mscroggs** [comment]
FEniCS/dolfinx#1369 needs a bit of attention still, but I can make a minor change in ffcx to allow this one to be merged without breaking dolfin

**Pull request merged**